### PR TITLE
Podman info add support for status of standard available cgroup controllers

### PR DESCRIPTION
**Pull request by flouthoc**
Following PR adds support for reflecting availability status of cgroup controllers to `info`.  Patch attempts to resolve https://github.com/containers/podman/issues/10306

### CODE REVIEW

File: libpod/define/info.go
Comment (rhatdan) [inline libpod/define/info.go:30]: I think this would be better in a subgroup,  CGroupInfo {
     CPUShares...
     MemoryLimit ...
     PidsLimit ...
}

And Alphabetized.

Comment (flouthoc) [reply libpod/define/info.go:30]: @rhatdan Sure makes sense, much cleaner i'll do the required changes.

File: pkg/cgroups/cgroups.go
Comment (AkihiroSuda) [inline pkg/cgroups/cgroups.go:136]: Not robust

https://github.com/kubernetes-sigs/kind/issues/2254#issuecomment-842888408

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

File: pkg/cgroups/cgroups.go
Comment (flouthoc) [reply pkg/cgroups/cgroups.go:136]: @AkihiroSuda ah I think i'll get prefix from group manager, that should resolve this. Thanks for the review :) i'll try to fix these.

File: libpod/info.go
Comment (AkihiroSuda) [inline libpod/info.go:112]: Wondering it may make more sense to just return `CgroupControllers []string` here.

(Docker-compatible REST API should emulate Docker REST API, though)

Comment (flouthoc) [reply libpod/info.go:112]: @AkihiroSuda `docker info` spits boolean for separate controllers so i tried keeping it same. But i guess a common string for all available controllers would be better. WDYT ?

File: pkg/cgroups/cgroups.go
Comment (giuseppe) [reply pkg/cgroups/cgroups.go:136]: can't we read those from the current cgroup so we don't have to worry about systemd/cgroupfs or rootless/rootful?

Comment (giuseppe) [reply pkg/cgroups/cgroups.go:136]: probably for root it still makes sense to read them from the root cgroup

File: libpod/info.go
Comment (AkihiroSuda) [reply libpod/info.go:112]: Yes, for Podman CLI, returning the common string slice SGTM.
Docker-compatible REST API should retain Docker-compatible booleans.

File: pkg/cgroups/cgroups.go
Comment (flouthoc) [reply pkg/cgroups/cgroups.go:136]: @giuseppe Yeah I think for rootless i could just get available controllers from current cgroup. That should be more standard.

File: libpod/info.go
Comment (flouthoc) [reply libpod/info.go]: @AkihiroSuda I tried this but dumping all the available controllers is making `info` too noisy, for instance on my machine even single liner has too much content. I guess docker only prints standard controllers maybe because this is too verbose but i am cool with this.
On my machine `info` shows this
```bash
cgroupControllers:
  - blkio
  - cpu
  - cpu,cpuacct
  - cpuacct
  - cpuset
  - devices
  - freezer
  - hugetlb
  - memory
  - net_cls
  - net_cls,net_prio
  - net_prio
  - perf_event
  - pids
  - rdma
  - systemd
  - unified
```

File: libpod/define/info.go
Comment (flouthoc) [reply libpod/define/info.go]: @rhatdan This is resolved. But format might change as there is a discussion going below for `CgroupControllers []string`

File: pkg/cgroups/cgroups.go
Comment (flouthoc) [reply pkg/cgroups/cgroups.go]: @AkihiroSuda @giuseppe PTAL for `rootless cgroupv2` now we are marking the controller availability as `false` if it got failed during `initialization` of cgroup. This seems to be more simple and wouldn't require handling caveats of systemd/cgroupfs unit name.

**Comment by rhatdan** [comment]
Please update the man page with the new output.

File: pkg/cgroups/cgroups.go
Comment (flouthoc) [reply pkg/cgroups/cgroups.go]: @AkihiroSuda @giuseppe PTAL made some changes I think current checking available controller for current user is better. `Service` and `scope` would anyways inherit this. Here is the output for `rootless` / `cgroupv2` from my machine.

```bash
host:
  arch: amd64
  buildahVersion: 1.20.2-dev
  cgroupInfo:
    cpuShares: false
    memoryLimit: true
    pidsLimit: true
  cgroupManager: systemd
  cgroupVersion: v2
```

**Comment by flouthoc** [comment]
@rhatdan Updated Man pages.

File: libpod/info.go
Comment (AkihiroSuda) [reply libpod/info.go]: For cgroup v1, the right way is to parse /proc/cgroups

Comment (flouthoc) [reply libpod/info.go]: @AkihiroSuda Would still give the same output as i have all controllers enabled on v1 
```bash
$ cat /proc/cgroups
#subsys_name	hierarchy	num_cgroups	enabled
cpuset	0	230	1
cpu	0	230	1
cpuacct	0	230	1
blkio	0	230	1
memory	0	230	1
devices	0	230	1
freezer	0	230	1
net_cls	0	230	1
perf_event	0	230	1
net_prio	0	230	1
hugetlb	0	230	1
pids	0	230	1
rdma	0	230	1
```

Comment (AkihiroSuda) [reply libpod/info.go]: not "same" :)

Comment (flouthoc) [reply libpod/info.go]: @AkihiroSuda ah i get it. So i guess logic for existing `getAvailableControllers` https://github.com/containers/podman/blob/master/pkg/cgroups/cgroups.go#L129 for cgroupv1 needs to be changed or i can add a new function but dont think redundant code for similar task is a good idea. @rhatdan @giuseppe @AkihiroSuda  If we can mutually agree upon having `CgroupControllers []string` in `info` then i think we can proceed with removing old logic in `master` and use `/proc/cgroups` for cgroupv1

File: pkg/cgroups/cgroups.go
Comment (AkihiroSuda) [inline pkg/cgroups/cgroups.go:145]: IIUC the controllers there are not always available for rootless users.
Reading the cgorup.subtree_control of Podman process itself might be better (it is still not robust, either, but at least better...)

Comment (AkihiroSuda) [reply pkg/cgroups/cgroups.go:145]: cc @giuseppe @kolyshkin

Comment (flouthoc) [reply pkg/cgroups/cgroups.go:145]: We cant use subtree created by podman process as it wouldn't exist , if `podman info` is called prior to creating any container.

Comment (flouthoc) [reply pkg/cgroups/cgroups.go:145]: also i think podman cleans up it entries as soon as running container exits. Not sure.

Comment (giuseppe) [reply pkg/cgroups/cgroups.go:145]: yes the problem is that some controllers are not enabled for the rootless user (and we don't really know what controllers are enabled until we do the equivalent of `systemd-run --scope --user` with systemd).  So I agree with @AkihiroSuda that reading the current enabled controllers is a best effort way of doing it, which should be fine in most cases

File: libpod/info.go
Comment (giuseppe) [reply libpod/info.go]: cgroupv1 is not really supported for rootless, even if manually chowned as the delegation is unsafe.

For root, we can parse `/proc/cgroups` or `/proc/self/cgroups` but for rootless we can just force every controller to be disabled.

File: pkg/cgroups/cgroups.go
Comment (flouthoc) [reply pkg/cgroups/cgroups.go]: @giuseppe @AkihiroSuda This is resolved in latest commit.

File: libpod/info.go
Comment (flouthoc) [reply libpod/info.go]: @giuseppe @AkihiroSuda Thank a lot . I have added changes, this is resolved in latest commit

File: libpod/define/info.go
Comment (flouthoc) [reply libpod/define/info.go]: @rhatdan Made some changes in latest commit , added `CgroupControllers []string` instead. Could you please review that.

**Comment by flouthoc** [comment]
@giuseppe @AkihiroSuda @rhatdan I have addressed all the comments in latest commit. Could you guys please take a look.

**Comment by flouthoc** [comment]
Missing test is a flake i think.

File: pkg/cgroups/cgroups.go
Comment (flouthoc) [reply pkg/cgroups/cgroups.go]: @giuseppe @AkihiroSuda Thank a lot . I have added changes, this is resolved in latest commit

**Comment by flouthoc** [comment]
It was a flake force push solved it. 😃

**Review by AkihiroSuda — APPROVED** [review]

**Comment by flouthoc** [comment]
yay!!! thanks @AkihiroSuda  waiting for @giuseppe and @rhatdan 's approval.

File: pkg/cgroups/cgroups.go
Comment (giuseppe) [inline pkg/cgroups/cgroups.go:145]: let's use `errors.Wrapf()` here

File: pkg/cgroups/cgroups.go
Comment (giuseppe) [inline pkg/cgroups/cgroups.go:133]: nit: extra space before "systemd"

**Comment by mheon** [comment]
Restarted a single flake.
/approve
LGTM

**Comment by flouthoc** [comment]
@mheon Thanks 😄

**Comment by flouthoc** [comment]
@mheon How do we restart single flake ? There is still one check which has a flake.

File: pkg/cgroups/cgroups.go
Comment (flouthoc) [reply pkg/cgroups/cgroups.go]: Resolved in latest commit.

File: pkg/cgroups/cgroups.go
Comment (flouthoc) [reply pkg/cgroups/cgroups.go]: Resolved in latest commit.

**Review by giuseppe — APPROVED** [review]
LGTM

**Comment by mheon** [comment]
/lgtm

@flouthoc I click the "Details" link on the test in question. The Github summary page it takes you to should have a "Retest" button if the test is red.

**Pull request merged**
